src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/CodeGeneration/CodeWriter.cs

```diff
@@ -65,17 +66,25 @@ private void AddTextChunk(ReadOnlyMemory<char> value)
         }
 
         // If we're at the start of a page, we need to add the page first.
-        var lastPage = _pageOffset == 0
-            ? _pages.AddLast(new ReadOnlyMemory<char>[PageSize]).Value
-            : _pages.Last!.Value;
+        ReadOnlyMemory<char>[] lastPage;
+
+        if (_pageOffset == 0)
+        {
+            lastPage = ArrayPool<ReadOnlyMemory<char>>.Shared.Rent(PageSize);
+            _pages.AddLast(lastPage);
+        }
+        else
+        {
+            lastPage = _pages.Last!.Value;
+        }
 
         // Add our chunk of text (the ReadOnlyMemory<char>) and increment the offset.
         lastPage[_pageOffset] = value;
         _pageOffset++;
 
         // We've reached the end of a page, so we reset the offset to 0.
         // This will cause a new page to be added next time.
-        if (_pageOffset == PageSize)
+        if (_pageOffset == lastPage.Length)
```

Comment:
Good change, though it's a bit subtle. Could you add a comment or two describing this behavior?

Reply:
Comment added, hopefully more obvious now.